prusti-contracts/prusti-contracts/src/core_spec/mem.rs

```diff
@@ -0,0 +1,65 @@
+use crate::*;
+
+#[extern_spec]
+mod core {
+    mod mem {
+        use crate::*;
+        
+        #[pure]
+        #[ghost_constraint(T: core_spec::mem::KnownSize, [ensures(result == T::size())])]
+        fn size_of<T>() -> usize;
+        
+        #[pure]
+        #[ghost_constraint(T: core_spec::mem::KnownSize, [ensures(result == T::align())])]
+        fn align_of<T>() -> usize;
+		
+		#[ensures(*x === old(snap(y)) && *y === old(snap(x)))]
+		fn swap<T>(x: &mut T, y: &mut T);
+    }
+}
+
+pub trait KnownSize {
+    #[pure]
+    fn size() -> usize;
+
+    #[pure]
+    fn align() -> usize;
+}
+
+macro_rules! known_size_spec {
+    ($t:ty, $size:expr, $align:expr) => {
+        #[refine_trait_spec]
+        impl KnownSize for $t {
+            #[pure]
+            #[ensures(result == $size)]
+            fn size() -> usize {
+                $size
+            }
+        
+            #[pure]
+            #[ensures(result == $align)]
+            fn align() -> usize {
+                $align
+            }
+        }
+    };
+}
+
+known_size_spec!(bool, 1, 1);
+
+known_size_spec!(i8, 1, 1);
+known_size_spec!(i16, 2, 2);
+known_size_spec!(i32, 4, 4);
+known_size_spec!(i64, 8, 8);
+known_size_spec!(i128, 16, 16);
+
+known_size_spec!(u8, 1, 1);
+known_size_spec!(u16, 2, 2);
+known_size_spec!(u32, 4, 4);
+known_size_spec!(u64, 8, 8);
+known_size_spec!(u128, 16, 16);
+
+known_size_spec!(f32, 4, 4);
+known_size_spec!(f64, 8, 8);
+
+// usize/isize are not specified exactly because they are platform-dependent and programmers should not depend on their specific values anyway.
```

Comment:
We could say that the value is one of the three possible for Rust, i.e. [16-, 32- or 64-bit](https://doc.rust-lang.org/src/core/num/mod.rs.html#243-268). However, I think we can do even better: put the `known_size_spec!` under a `cfg`, so when compiling the correct value is chosen.

Reply:
Yeah, I considered something like that. This would allow you to ensure that your code is correct across different word sizes without having to run Prusti for each platform, while specifying the exact size gives information the programmer should maybe not be relying on? But I can absolutely specify it if you think it makes sense.

Reply:
I think we make the same assumption (that the verification machine = the target machine) when processing MIR [`SizeOf`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.NullOp.html#variant.SizeOf) operations. So for now, I would suggest using the `cfg` approach to be consistent. In the future we might think about "cross-verification".